### PR TITLE
feat: support downloading from customized url

### LINKED ISSUE #24: Support downloading ORAS through mirror
In airgapped situations, users will need to download the ORAS binary from a mirrored location.

### PR DESCRIPTION
**Pull request by wangxiaoxuan273**
resolves #24

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -56,3 +56,54 @@ jobs:
...
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v3
+
+    - name: Setup ORAS using URL
+      uses: ./
+      with:
+        url: https://github.com/oras-project/oras/releases/download/v1.0.0/oras_1.0.0_linux_amd64.tar.gz
+        checksum: "8533c9ea1e5a0d5eb1dfc5094c0e8ef106d15462f8a119077548f88937ed2133"
```
Comment (qweeah) [inline .github/workflows/test.yml:71]: What if someday we remove 1.0.0? To reduce long-term maintenance effort, it's better to grab an existing version in `release.json`. E.g.

```bash
jq -r 'keys_unsorted[0] as $k | .[$k].linux.amd64' src/lib/data/releases.json
```

File: src/lib/release.ts [hunk 2]
```diff
@@ -27,7 +27,15 @@ interface releases {
...
 // Get release info of a certain verion of ORAS CLI
-export function getReleaseInfo(version: string) {
+export function getReleaseInfo(version: string, url: string, checksum: string) {
+    // if customized ORAS CLI link is provided
+    if (url && checksum) {
+        return {
+            checksum: checksum, 
+            url: url
+        } 
+    }
```
Comment (shizhMSFT) [inline src/lib/release.ts:37]: What if `url` is provided but `checksum` is not provided? What's the proper error message?

File: src/setup.ts [hunk 3]
```diff
@@ -21,21 +21,28 @@ async function setup(): Promise<void> {
   try {
     // inputs from user
     const version: string = core.getInput('version');
+    const url: string = core.getInput('url');
+    const checksum = core.getInput('checksum').toLowerCase();
+
+    // sanity check
+    if (url && !checksum) {
+      throw new Error("user provided url of customized ORAS CLI release but without SHA256 checksum");
+    }
```
Comment (shizhMSFT) [inline src/setup.ts:30]: Should this check be done in `getReleaseInfo`?

File: src/lib/release.ts (hunk 2 above)
Comment (shizhMSFT) [reply src/lib/release.ts:37]: What will happen when `checksum` is provided if `version` is provided but `url` is not provided?

Comment (shizhMSFT) [reply src/lib/release.ts:37]: What if all of three parameters are provided?

File: src/setup.ts (hunk 3 above)
Comment (wangxiaoxuan273) [reply src/setup.ts:30]: move it into `getReleaseInfo`

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -56,3 +56,68 @@ jobs:
...
+    outputs:
+      url: ${{ steps.get-url.outputs.URL }}
+      checksum: ${{ steps.get-checksum.outputs.CHECKSUM }}
+    steps:
+      - id: checkout
+        uses: actions/checkout@v3
+      - id: get-url
+        run: echo "URL=$(jq -r 'keys_unsorted[0] as $k | .[$k].linux.amd64.url' src/lib/data/releases.json)" >> "$GITHUB_OUTPUT"
+      - id: get-checksum
+        run: echo "CHECKSUM=$(jq -r 'keys_unsorted[0] as $k | .[$k].linux.amd64.checksum' src/lib/data/releases.json)" >> "$GITHUB_OUTPUT"
```
Comment (qweeah) [inline .github/workflows/test.yml:71]: Suggest to read the file only once and merge the steps
```suggestion
      - id: get-checksum-url
        run: |
        RELEASE=$(jq -r 'keys_unsorted[0] as $k | .[$k].linux.amd64' src/lib/data/releases.json)
        echo "CHECKSUM=$(echo $RELEASE | jq -r '.checksum')" >> "$GITHUB_OUTPUT"
        echo "URL=$(echo $RELEASE | jq -r '.url')" >> "$GITHUB_OUTPUT"
```

File: dist/index.js [hunk 5]
```diff
@@ -6691,7 +6691,22 @@ exports.getBinaryExtension = exports.mapArch = exports.mapPlatform = exports.get
 const os = __importStar(__nccwpck_require__(2037));
 const releases_json_1 = __importDefault(__nccwpck_require__(2387));
 // Get release info of a certain verion of ORAS CLI
-function getReleaseInfo(version) {
+function getReleaseInfo(version, url, checksum) {
+    // if customized ORAS CLI link and checksum are provided, version is ignored
+    if (url && checksum) {
+        return {
```
Comment (qweeah) [inline dist/index.js:6697]: Better place the comment in the corresponding code path.
```suggestion
    if (url && checksum) {
        // if customized ORAS CLI link and checksum are provided, version is ignored
        return {
```

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -56,3 +56,68 @@ jobs:
...
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v3
+
+    - name: Setup ORAS using URL
+      uses: ./
+      with:
+        url: ${{ needs.create-test-variables.outputs.url }}
+        checksum: ${{ needs.create-test-variables.outputs.checksum }}
+
```
Comment (qweeah) [inline .github/workflows/test.yml:86]: Need a test on setup  ORAS using checksum without URL

Comment (wangxiaoxuan273) [reply .github/workflows/test.yml:87]: added a test

**Review by qweeah — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**